File: Study5 - Tree/Week31/Day215/boj12888_hsk.py
#같은 로직으로 nodejs에서는 실패가 떠서 python으로 작성하였습니다.
#js 실패 이유 : js에서는  Number.MAX_SAFE_INTEGER ( = 9007199254740991 ) 보다 큰 정수는
#오차가 있을 수 있기 때문에 정확하지 않은 값이 산출됩니다.
## MAX_SAFE_INTEGER 라는 개념이 있다니 신기한 언어네요.. 수고하셨습니다
import sys
input = sys.stdin.readline

H = int(input())
numOfCarNeeded = [1] * (H + 2)
numOfCarNeeded[0] = 1
numOfCarNeeded[1] = 1

## 이런 규칙도 있군요 : 22 규칙을 잘 찾으신 것 같아요!
for i in range(2, H + 1):
    if i % 2 == 0:
        numOfCarNeeded[i] = 2 * numOfCarNeeded[i - 1] + 1;
    else:
        numOfCarNeeded[i] = 2 * numOfCarNeeded[i - 1] - 1;

print(numOfCarNeeded[H])

File: Study5 - Tree/Week31/Day215/boj12888_hsk.js
const fs = require('fs');
const input = (process.platform === 'linux' ? fs.readFileSync('/dev/stdin').toString().trim() : `60`).split('\n');
const H = BigInt(input[0]);
const numOfCarNeeded = [BigInt(1), BigInt(1)];

for (let i = 2; i <= H; i++) {
  if (i % 2 === 0) {
    numOfCarNeeded[i] = BigInt(2) * numOfCarNeeded[i - 1] + BigInt(1);
  } else {
    numOfCarNeeded[i] = BigInt(2) * numOfCarNeeded[i - 1] - BigInt(1);
  }
}

console.log(numOfCarNeeded[H].toString());
// MAX_SAFE_INTEGER보다 큰 값들은 BigInt로 계산을 하고 숫자를 문자열로 바꿔서 제출하여 성공하였습니다.
